akka-javasdk/src/main/scala/akka/javasdk/impl/agent/AgentImpl.scala

```diff
@@ -358,31 +353,9 @@ private[impl] final class AgentImpl[A <: Agent](
 
   override def callTool(request: SpiAgent.ToolCallRequest): Future[String] = {
 
-    val toolInvoker =
-      functionTools.getOrElse(request.name, throw new IllegalArgumentException(s"Unknown tool ${request.name}"))
-
-    val mapper = serializer.objectMapper
-    val jsonNode = mapper.readTree(request.arguments)
-
-    val methodInput =
-      toolInvoker.paramNames.zipWithIndex.map { case (name, index) =>
-        // assume that the paramName in the method matches a node from the json 'content'
-        val node = jsonNode.get(name)
-        val typ = toolInvoker.types(index)
-        val javaType = mapper.getTypeFactory.constructType(typ)
-        mapper.treeToValue(node, javaType).asInstanceOf[Any]
-      }
-
-    Future {
-      val toolResult = toolInvoker.invoke(methodInput)
-
-      if (toolInvoker.returnType == Void.TYPE)
-        "SUCCESS"
-      else if (toolInvoker.returnType == classOf[String])
-        toolResult.asInstanceOf[String]
-      else
-        mapper.writeValueAsString(toolResult)
-    }(sdkExecutionContext)
+    val metadata = MetadataImpl.of(request.metadata)
+    val agentContext = new AgentContextImpl(sessionId, regionInfo.selfRegion, metadata, request.span, tracerFactory)
+    router.invokeTool(request, agentContext)(sdkExecutionContext)
```

Comment:
I completely rebuilt tool invocation, because in the previous version, it was not possible to pass metadata and/or span to have tracing at this level as well.

Reply:
This is looking fine, but I have a concern. 

As suggested today in the stand-up, we plan to move the `callTool` method out of the SpiAgent and make it part of the Effect. 

Then this might not be necessary anymore. 
Anyway, let's move forward with what we have here and we refactor again if needed.